Mobile: confirm encryption password

nathanleiby opened the pull request:
Issue: #1656

Same idea as #1936 except for Mobile

--

_Before_

<img width="371" alt="iPhone_X_—_12_2" src="https://user-images.githubusercontent.com/102242/66025777-1e937980-e4ac-11e9-8cd8-7e7aeb3e1b29.png">

_New UI:_

(Verified that it sets up encryption as normal if passwords match)

<img width="366" alt="iPhone_X_—_12_2" src="https://user-images.githubusercontent.com/102242/66025417-42a28b00-e4ab-11e9-932e-50b591a3020b.png">

**Error cases**

_Errors if passwords don't match_

<img width="370" alt="iPhone_X_—_12_2" src="https://user-images.githubusercontent.com/102242/66025582-acbb3000-e4ab-11e9-9f62-fc1383cc038e.png">

_Errors if either field is empty and you click "enable"_

<img width="364" alt="iPhone_X_—_12_2" src="https://user-images.githubusercontent.com/102242/66025479-6fef3900-e4ab-11e9-9e3c-af3c1ab17431.png">

<img width="379" alt="iPhone_X_—_12_2" src="https://user-images.githubusercontent.com/102242/66025524-8b5a4400-e4ab-11e9-8e91-30f91fb330e5.png">

laurent22 commented on ReactNativeClient/lib/components/screens/encryption-config.js:
```diff
@@ -131,6 +132,9 @@ class EncryptionConfigScreenComponent extends BaseScreenComponent {
 				if (!password) throw new Error(_('Password cannot be empty'));
+				const password2 = this.state.passwordPromptConfirmAnswer;
+				if (!password2) throw new Error(_('Confirm password cannot be empty'));
+				if (password !== password2) throw new Error(_('Passwords do not match'));
```
As those are translated strings, please use the same text as on your CLI update: "Passwords did not match!"

laurent22 commented on ReactNativeClient/lib/components/screens/encryption-config.js:
```diff
@@ -141,6 +145,7 @@ class EncryptionConfigScreenComponent extends BaseScreenComponent {
 		return (
 			<View style={{ flex: 1, borderColor: theme.dividerColor, borderWidth: 1, padding: 10, marginTop: 10, marginBottom: 10 }}>
 				<Text style={{ fontSize: theme.fontSize, color: theme.color }}>{_('Enabling encryption means *all* your notes and attachments are going to be re-synchronised and sent encrypted to the sync target. Do not lose the password as, for security purposes, this will be the *only* way to decrypt the data! To enable encryption, please enter your password below.')}</Text>
+				<Text>Password: </Text>
```
Translatable string: `{_('Password:')}`

laurent22 commented on ReactNativeClient/lib/components/screens/encryption-config.js:
```diff
@@ -150,6 +155,16 @@ class EncryptionConfigScreenComponent extends BaseScreenComponent {
 						this.setState({ passwordPromptAnswer: text });
 					}}
 				></TextInput>
+				<Text>Confirm Password: </Text>
```
Translatable string: `{_('Confirm password:')}`

laurent22 commented on ReactNativeClient/lib/components/screens/encryption-config.js:
```diff
@@ -150,6 +155,16 @@ class EncryptionConfigScreenComponent extends BaseScreenComponent {
+				<Text>Confirm Password: </Text>
+				<TextInput
+					selectionColor={theme.textSelectionColor}
+					style={{ margin: 10, color: theme.color, borderWidth: 1, borderColor: theme.dividerColor }}
```
Please move the style to a variable so that it can be reused by this component and the previous one.

laurent22 commented:
Looks good overall except for the minor issues I mentioned in comment. It will need to be tested before merging it though.

tessus commented:
I can test as soon as it is ready. @nathanleiby just give me a shout, when you need it tested.

nathanleiby commented:
Thanks for feedback! Made the changes suggested. https://github.com/laurent22/joplin/pull/1939/commits/342c97f33b299f009a620a696f3ff5c0177e175e 

I also explored adding labels to the `TextInput`s.. looks like they don't support labels, so I added placeholder text (https://facebook.github.io/react-native/docs/textinput). LMK if you (dis)like.

Also: Got iOS simulator running! I updated the PR description with various screenshots.

tessus commented:
2 comments:

- The text should be `Passwords do not match!` (past tense is never used in this situation) I missed that part in the CLI review. but this is an easy fix.
- what happens, if both fields are empty?

I haven't tested it myself yet, so I guess the second item will reveal itself as soon as I do.

But I'm heading to bed now. I have to get up in a few hours.

tessus commented on ReactNativeClient/lib/components/screens/encryption-config.js:
```diff
@@ -131,6 +138,9 @@ class EncryptionConfigScreenComponent extends BaseScreenComponent {
 				if (!password) throw new Error(_('Password cannot be empty'));
+				const password2 = this.state.passwordPromptConfirmAnswer;
+				if (!password2) throw new Error(_('Confirm password cannot be empty'));
+				if (password !== password2) throw new Error(_('Passwords did not match!'));
```
Please change `did` to `do`.

tessus replied:
Hmm, it shows 3 lines, instead of the one I clicked. So my previous comment is not very clear.
Please change: `Passwords did not match!` to `Passwords do not match!`

laurent22 replied:
Also please change it in the CLI app (if you rebase, you'll get your latest changes).

tessus replied:
@laurent22 no longer possible, since it is already merged. I just made the change for CLI myself 74c2bbc2

tessus replied:
@laurent22 ah, you meant in this PR. yep, that would have worked ;-) haha, my bad. I fixed it already for the CLI app.

tessus commented:
@nathanleiby I can also make the change for you, if you don't have the time.

nathanleiby commented:
@tessus Apologies for the delay. Updated the text in:https://github.com/laurent22/joplin/pull/1939/commits/0acc5b525c1ba579e41b88eb74714d9623844092


For future reference, I'm totally fine if you make text and other changes that allow a merge. Thanks!

tessus commented:
Thanks for the update!

laurent22 commented:
@tessus, have you had the opportunity to build and test this PR? If so please go ahead and merge. Otherwise, I'll test it on my side, just let me know.

tessus commented:
@laurent22 Yes, I have tested it on iOS and Android.
